Dictionaries

rodrigo-pino opened the pull request:
Fixes #63 

Adds a new chapter explaining `DictFelt252<T>`.

Content wise:
- [x] Basic Dict functionality
- [x] Dictionaries Underneath
  - [x] Explain Entry List
  - [x] Explain squashing and why it is need
  - [x] Explain Destruct
- [x] Additional example use of Dicts
  - [x] Using the `entry` method
  - [x] How dictionaries interact with Structs
  - [ ] ~How dictionaries interact with Enums?~

Also it still lacks polishing, specially the later sections, and run it through a spell check.

enitrat commented:
> How dictionaries interact with Structs
> How dictionaries interact with Enums?

I'm not sure about Enums, but for structs, you can create a Felt252Dict<Nullable<T>> to store your struct in the dict easily!

enitrat commented on src/ch15-01-dictionaries.md:
````diff
@@ -0,0 +1,204 @@
+The use of this entry list shows how there isn't any rewriting, just the creation of new memory cells per `Felt252Dict<T>` interaction. Let's show an example of this using the `balances` dictionary from the previous section and inserting the users 'Alex' and 'Maria':
+
+```rust
+balances.insert('Alex', 100_u64);
````
To reference this extract of cairo, you can create a `.cairo` file under listings and use the anchor system to import it here: https://rust-lang.github.io/mdBook/format/mdbook.html?highlight=ANCHOR#including-portions-of-a-file

enitrat commented on src/ch15-01-dictionaries.md:
```diff
@@ -0,0 +1,204 @@
+
+Notice that since 'Alex' was inserted twice, it appears twice and the `previous` and `current` values are set properly. Also reading from 'Maria' registered an entry with no change from previous to current values.
+
+This approach to implement `Felt252Dict<T>` means that for each reading/writing operations there is a scan for the whole entry list in search the last entry with the same `key`. Once the entry have been found, its `new_value` is extacted and used on the new entry to be added as the `previous_value`. This means that interacting with `Felt252Dict<T>` has a worst time complexity of `O(n)` where `n`  ...
```
> This means that interacting with `Felt252Dict<T>` has a worst time complexity of `O(n)` where `n` is the amount of entries in the list.

Are we sure about this? How can we evaluate that?

rodrigo-pino replied:
Uff I gather that after diving into the sierra to CASM compiler, and seen the generated code. It was a loop through the memory segment comparing keys until the last one. It was some time a go, I'll recheck and  try to link the file where I saw it

enitrat replied:
Awesome, thanks! 🙌

enitrat commented on src/ch15-01-dictionaries.md:
```diff
@@ -0,0 +1,204 @@
+
+If you ran the examples from [Basic use of Dictionaries](##Basic use of Dictionaries) you'd probably notice that you never did a call to squash dictionary, but program compiled succesfully nonetheless. What happened behind the scene was that squash was called automatically via the `Felt252Dict<T>` implementation of the `Destruct<T>` trait. This call ocurred just before `balance` went out of scope ...
+
+The `Destruct<T>` trait represents another way of removing instances out of scope apart from `Drop<T>`. The main difference between these two is that `Drop<T>` is treated as a no-op operation, meaning it does not generate new CASM while `Destruct<T>` does not have this restriction. The only type which actively uses the `Destruct<T>` trait is `Felt252Dict<T>`, for every other type `Destruct<T>` an ...
```
Can you add a reference to appendix-03 ### Drop and Destruct?

enitrat commented on src/ch15-01-dictionaries.md:
```diff
@@ -0,0 +1,204 @@
+
+Remember when we mentioned
+
+by using the entry function you can update the dictionary in a new way
```
quotes

rodrigo-pino replied:
This was the part that I hadn't wrote 🙈 , just random ideas to link later. The remaining comments address unwritten stuff as well

enitrat commented on src/ch15-01-dictionaries.md:
```diff
@@ -0,0 +1,204 @@
+by using the entry function you can update the dictionary in a new way
+
+It returns a dict entry according to a key
+
```
extra newline

enitrat commented on src/ch15-01-dictionaries.md:
```diff
@@ -0,0 +1,204 @@
+Once you do such entry.finalize you are given back access to the normal dictionary
+
+There is no performance gain, but it represents another idiomatic way of modifying a dictionary.
+
```
Can you show an example for this?

enitrat reviewed (commented):
Incredible work @rodrigo-pino, it's super clear and well written 🔥

rodrigo-pino commented:
I added the core of what I wanted to say in the last section. On monday I'll give it another pass to improve coherence and typo correction. Also to use the listings for the code.

I wanted to reference the Nullable in the book, but I don't think it is written yet, so I added a small paragraph explaining it but it may be incomplete, or not entirely correct since I haven't researched it in depth as for now.

rodrigo-pino commented on src/ch15-01-dictionaries.md:
```diff
@@ -0,0 +1,204 @@
+
+## Dictionaries Underneath
+
+One of the constraints of being a non-deterministic language is that the memory system is inmutable, so in order to simulate mutability, the team decided to implement `Felt252Dict<T>` as list of entries. Each of the entries represent a time where a dictionary was accessed for reading/updating/writing purposes. An entry have three fields:
```
👍

enitrat reviewed (approved):
Awesome!!!
